### PR TITLE
Dummy task

### PR DESCRIPTION
**Pull request by inexxt**
This PR: 
 - creates dummy task, using old dummy task concept (PoW computations)
 - in particular, dummy task makes it easier to handle additional code and data 
 - turns off old dummy task tests
 - refactors `Task`s structure a little, creating some interfaces and allowing for easier task creation
 - annotates types in a few places
 - adds `enforce` dependency, for strict types checking (it is not enabled by default, and currently only works in the `DummyTask` class, but because of type annotations can easily be used everywhere)
 - <del> removes randomness from luxrender tests</del> done in #1286

Known bugs:
 - During task creation, temporary directory is created under /tmp (or equivalent), which is not removed aftrewards. Partial solution to this will be creation of the directory under temp inside task dir, but it has to be done in the Task itself, not in the TaskDefinition.

### CODE REVIEW

File: apps/dummy/task/dummytaskstate.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        # but for now TaskDefinition doesn't know root_path
+        # task_root_path = ""
+        # self.tmp_dir = DirManager().get_task_temporary_dir(self.task_id, True)
+
+        self.tmp_dir = tempfile.mkdtemp()
+
+        self.shared_data_files = list(self.resources)
+        self.code_files = ls_R(self.code_dir)
+
+        os.symlink(self.code_dir, os.path.join(self.tmp_dir, "code"))
```
Comment (badb) [inline apps/dummy/task/dummytaskstate.py]: Throws error on Windows when called by unprivileged user. You'll probably need some symlink_or_copy method here.

File: .gitignore [hunk 2]
```diff
@@ -70,3 +70,7 @@ Installer/Installer_Win/golem-*
 
 # virtualenv
 venv/*
+
+# luxrender test files
+luxrender-output.exr
```
Comment (badb) [inline .gitignore]: Are those files left by some golem tests?

Comment (inexxt) [reply .gitignore]: They were, I think it's fixed now

**Comment by inexxt** [comment]
I think it will be necessary to turn off linters this time (or turn them on only for specific files) - because a lot of code is touched in this PR, linters found ~2k warnings and errors.

File: apps/dummy/task/dummytaskstate.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        self.code_files = ls_R(self.code_dir)
+
+        symlink_or_copy(self.code_dir, os.path.join(self.tmp_dir, "code"))
+
+        # makes sense when len(..) > 1
+        # common_data_path = os.path.commonpath(self.shared_data_files)
+        # but we only have 1 file here
+        common_data_path = os.path.dirname(list(self.shared_data_files)[0])
+
+        os.symlink(common_data_path, os.path.join(self.tmp_dir, "data"))
```
Comment (badb) [inline apps/dummy/task/dummytaskstate.py]: Same problem as above, should be symlink_or_copy

Comment (inexxt) [reply apps/dummy/task/dummytaskstate.py]: Fixed

File: tests/golem/logs/golemdocker_test.log.2017-08-09 [hunk 4]
```diff
@@ -0,0 +1,18 @@
+2017-08-09 17:08:47 WARNING  golem.appconfig                     Can't open file /home/jacek/.local/share/golem/SimpleEnv/GOODSETTINGS: [Errno 2] No such file or directory: '/home/jacek/.local/share/golem/SimpleEnv/GOODSETTINGS'
```
Comment (badb) [inline tests/golem/logs/golemdocker_test.log.2017-08-09]: Are these logs needed?

File: tests/apps/dummy/task/test_dummytaskstate.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    def test_add_to_resources(self):
+        td = DummyTaskDefinition(DummyTaskDefaults())
+        td.resources = {os.path.join(get_golem_path(), "apps", "dummy", "test_data", "in.data")}
+        assert os.path.isfile(list(td.resources)[0])
+        with patch("tempfile.mkdtemp", lambda: self.tempdir):
+            td.add_to_resources()
+            assert os.path.isdir(td.tmp_dir)
+            assert isinstance(td.resources, set)
+            assert td.tmp_dir == self.tempdir
+            assert os.path.islink(os.path.join(td.tmp_dir, "code"))
```
Comment (badb) [inline tests/apps/dummy/task/test_dummytaskstate.py]: It will fail on Windows without privilieged mode

Comment (inexxt) [reply tests/apps/dummy/task/test_dummytaskstate.py]: Fixed, changed to isdir

File: tests/golem/logs/golemdocker_test.log.2017-08-09 (hunk 4 above)
Comment (inexxt) [reply tests/golem/logs/golemdocker_test.log.2017-08-09]: No, a mistake with commit. All logs deleted now.

File: apps/core/task/coretask.py [hunk 6]
```diff
@@ -172,7 +213,23 @@ def computation_finished(self, subtask_id, task_result, result_type=0):
...
+        if subtask.get("status", None) == SubtaskStatus.finished:
+            # logger.warning("Subtask %r already accepted", subtask_id)
+            raise Exception("Subtask {} already accepted".format(subtask_id))
+        if subtask.get("status", None) not in [SubtaskStatus.starting,
+                                               SubtaskStatus.downloading,
+                                               SubtaskStatus.resent,
+                                               SubtaskStatus.finished,
+                                               SubtaskStatus.failure,
+                                               SubtaskStatus.restarted]:
+            # logger.warning("Subtask %r has wrong type", subtask_id)
```
Comment (banasiakadam60) [inline apps/core/task/coretask.py:229]: Please remove this 3 commented logs

File: apps/dummy/benchmark/benchmark.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    def __init__(self):
+        self._normalization_constant = 1000  # TODO tweak that
+        self.dummy_task_path = join(get_golem_path(),
+                                    "apps", "dummy", "test_data")
+
+        td = self._task_definition = DummyTaskDefinition(DummyTaskDefaults())
+        td.shared_data_files = [join(self.dummy_task_path, x) for x in
+                                td.shared_data_files]
+        # td.out_file_basename = join(tempfile.gettempdir(),
+        # td.out_file_basename)
```
Comment (banasiakadam60) [inline apps/dummy/benchmark/benchmark.py]: Remove or uncomment

File: apps/dummy/benchmark/benchmark.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        td.out_file_basename = td.out_file_basename
+
+        td.task_id = u"{}".format("dummy_benchmark")
```
Comment (banasiakadam60) [inline apps/dummy/benchmark/benchmark.py]: this:
`td.task_id = u"{}".format("dummy_benchmark")`
is equivalent to to:
`td.task_id = "dummy_benchmark"`
Please change it

File: apps/dummy/gui/controller/dummytaskcustomizer.py [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import logging
+
+from gui.controller.customizer import Customizer
+
+logger = logging.getLogger("apps.dummy")
```
Comment (banasiakadam60) [inline apps/dummy/gui/controller/dummytaskcustomizer.py:5]: Why do you need logger here? If I'm right, it isn't used

File: apps/dummy/resources/code_dir/computing.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    4 leftmost bytes of H is less or equal difficulty.
+    :param str data_file: file with shared task data
+    :param str subtask_string: subtask-specific part of data
+    :param int difficulty: required difficulty
+    :param int result_size: size of the solution string S
+    :rtype DummyTaskResult\
+    """
+    print('[DUMMY TASK] computation started, data_file = ', data_file,
+          ', result_size = ', result_size,
+          ', difficulty = 0x%08x' % difficulty)  # TODO remove that print
```
Comment (banasiakadam60) [inline apps/dummy/resources/code_dir/computing.py]: @TODO?

File: apps/dummy/resources/code_dir/computing.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    t0 = time.clock()
+
+    with open(data_file, 'r') as f:
+        shared_input = f.read()
+    solution = find_pow(shared_input + subtask_string, difficulty, result_size)
+    result = '%x' % solution
+    assert len(result) == result_size
+
+    print('[DUMMY TASK] computation finished, time =', time.clock() - t0, 'sec')
```
Comment (banasiakadam60) [inline apps/dummy/resources/code_dir/computing.py]: Should it also be removed?

File: apps/dummy/resources/scripts/docker_dummytask.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
+from __future__ import print_function
+
+import imp
+import os
+
+import params  # This module is generated before this script is run
+
+OUTPUT_DIR = "/golem/output"
+WORK_DIR = "/golem/work"  # we don't need that, all the work is done in memory
```
Comment (banasiakadam60) [inline apps/dummy/resources/scripts/docker_dummytask.py:9]: It isn't used. Should it be removed?

File: tests/apps/lux/task/test_luxrendertask.py [hunk 13]
```diff
@@ -243,7 +244,7 @@ def test_get_merge_ctd_error(self, find_task_script_mock):
         assert any("Cannot find merger script" in log for log in l.output)
 
     def test_update_preview_with_exr(self):
-        p = Path(__file__).parent.parent.parent /\
+        p = Path(__file__).parent.parent.parent / \
```
Comment (banasiakadam60) [inline tests/apps/lux/task/test_luxrendertask.py]: IMO more readable will be:
```
        from golem.core.common import get_golem_path
        from os.path import join
        p = join(get_golem_path(), 'tests', "apps",
                  "rendering", "resources", "testfile.EXR")
```

But it's just a suggestion

File: tests/golem/docker/test_docker_dummy_task.py [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+        self.error_msg = None
+        self.dirs_to_remove = []
+        self.files_to_remove = []
+        self.node = None
+        self._send_task_failed = TaskServer.send_task_failed
+
+    @classmethod
+    def setUpClass(cls):
+        # TempDirFixture.setUpClass()
+        # DockerTestCase.setUpClass()
```
Comment (banasiakadam60) [inline tests/golem/docker/test_docker_dummy_task.py]: Remove comments

File: tests/golem/docker/test_docker_dummy_task.py [hunk 15]
```diff
@@ -0,0 +1,298 @@
...
+        TempDirFixture.tearDown(self)
+
+    @classmethod
+    def tearDownClass(cls):
+        rmlink_or_rmtree(cls.code_link)
+        rmlink_or_rmtree(cls.data_link)
+        os.rmdir(cls.test_tmp)
+        
+        # TempDirFixture.tearDownClass()
+        # DockerTestCase.tearDownClass()
```
Comment (banasiakadam60) [inline tests/golem/docker/test_docker_dummy_task.py]: As above

**Review by banasiakadam60 — CHANGES_REQUESTED** [review]

File: apps/dummy/resources/scripts/docker_dummytask.py (hunk 12 above)
Comment (inexxt) [reply apps/dummy/resources/scripts/docker_dummytask.py:9]: No, I left it here for "educational" purposes (so if someone is swapping his own code in place of `computation.py` and is trying to modify the docker script, he will notice that there is this `WORK_DIR` waiting for him to be used)

File: tests/apps/lux/task/test_luxrendertask.py (hunk 13 above)
Comment (inexxt) [reply tests/apps/lux/task/test_luxrendertask.py]: Done

File: apps/dummy/resources/scripts/docker_dummytask.py (hunk 12 above)
Comment (banasiakadam60) [reply apps/dummy/resources/scripts/docker_dummytask.py:9]: OK

File: apps/dummy/resources/code_dir/computing.py (hunk 11 above)
Comment (inexxt) [reply apps/dummy/resources/code_dir/computing.py]: TODO was actually about removing old-style print and using new-style print()

File: apps/dummy/resources/code_dir/computing.py (hunk 10 above)
Comment (inexxt) [reply apps/dummy/resources/code_dir/computing.py]: TODO was actually about removing old-style `print` and replacing with new style `print()`

File: apps/dummy/gui/controller/dummytaskcustomizer.py (hunk 9 above)
Comment (inexxt) [reply apps/dummy/gui/controller/dummytaskcustomizer.py:5]: Same as below, "educational" purposes

File: apps/dummy/benchmark/benchmark.py (hunk 8 above)
Comment (inexxt) [reply apps/dummy/benchmark/benchmark.py]: Done

File: apps/dummy/benchmark/benchmark.py (hunk 7 above)
Comment (inexxt) [reply apps/dummy/benchmark/benchmark.py]: Done

File: apps/core/task/coretask.py (hunk 6 above)
Comment (inexxt) [reply apps/core/task/coretask.py:229]: I didn't know if the function should `raise Exception` or just log the event and go on.
What do you think?
@badb ?

File: apps/dummy/resources/code_dir/computing.py (hunk 11 above)
Comment (inexxt) [reply apps/dummy/resources/code_dir/computing.py]: Same as above

File: tests/golem/docker/test_docker_dummy_task.py (hunk 14 above)
Comment (inexxt) [reply tests/golem/docker/test_docker_dummy_task.py]: Done

File: tests/golem/docker/test_docker_dummy_task.py (hunk 15 above)
Comment (inexxt) [reply tests/golem/docker/test_docker_dummy_task.py]: Done

File: apps/core/task/coretask.py (hunk 6 above)
Comment (badb) [reply apps/core/task/coretask.py:229]: This function should be called only if the subtasks was really accepted, ie. it was all right. This exception should never be triggered with normal execution, ie. keeping Exception should be better.

File: apps/core/task/coretask.py [hunk 16]
```diff
@@ -82,45 +86,80 @@ def _preview_result(result, single=False):
 
 
 class CoreTask(Task):
+    VERIFICATOR_CLASS = CoreVerificator  # type: Type[CoreVerificator]
+
+    # TODO maybe @abstract @property?
+    ENVIRONMENT_CLASS = None  # type: Type[Environment]
```
Comment (badb) [inline apps/core/task/coretask.py:92]: Or maybe default Enironment. Can we keep it with # TODO annotation for now?

File: apps/core/task/coretask.py [hunk 17]
```diff
@@ -295,24 +368,26 @@ def result_incoming(self, subtask_id):
         self.counting_nodes[self.subtasks_given[subtask_id]['node_id']].finish()
         self.subtasks_given[subtask_id]['status'] = SubtaskStatus.downloading
 
-    def query_extra_data_for_test_task(self):
-        return None  # Implement in derived methods
+    # TODO why is it here and not in the Task?
+    @abc.abstractmethod
+    def query_extra_data_for_test_task(self) -> ComputeTaskDef:
```
Comment (badb) [inline apps/core/task/coretask.py:373]: History... Should be moved to Task

File: apps/dummy/resources/scripts/docker_dummytask.py [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+
+    data_file = os.path.join(RESOURCES_DIR, "data", data_files[0])
+    result_path = os.path.join(OUTPUT_DIR, result_file)
+
+    solution = computing.run_dummy_task(data_file,
+                                        subtask_data,
+                                        difficulty,
+                                        result_size)
+
+    with open(result_path, "w") as f:  # TODO try catch and log errors
```
Comment (badb) [inline apps/dummy/resources/scripts/docker_dummytask.py:25]: https://github.com/golemfactory/golem/issues/1353

File: golem/task/tasktester.py [hunk 19]
```diff
@@ -10,7 +11,9 @@ class TaskTester(LocalComputer):
     TESTER_WARNING = "Task not tested properly"
     TESTER_SUCCESS = "Test task computation success!"
 
-    def __init__(self, task, root_path, success_callback, error_callback):
+    # TODO I think there should be Task, not CoreTask type
```
Comment (badb) [inline golem/task/tasktester.py:14]: https://github.com/golemfactory/golem/issues/1355

File: tests/apps/dummy/gui/controller/test_dummytaskcustomizer.py [hunk 20]
```diff
@@ -0,0 +1,24 @@
+from mock import Mock
+
+from golem.tools.assertlogs import LogTestCase
+from golem.tools.testdirfixture import TestDirFixture
+from gui.application import Gui
+from gui.applicationlogic import GuiApplicationLogic
+from gui.view.appmainwindow import AppMainWindow
+
+
+# TODO
```
Comment (badb) [inline tests/apps/dummy/gui/controller/test_dummytaskcustomizer.py:10]: Will be removed with removing GUI part

File: apps/dummy/benchmark/benchmark.py [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+    DummyTaskDefaults
+from apps.dummy.task.verificator import DummyTaskVerificator
+from golem.core.common import get_golem_path
+
+APP_DIR = join(get_golem_path(), 'apps', 'dummy')
+
+
+class DummyTaskBenchmark(CoreBenchmark):
+    def __init__(self):
+        self._normalization_constant = 1000  # TODO tweak that
```
Comment (badb) [inline apps/dummy/benchmark/benchmark.py:17]: https://github.com/golemfactory/golem/issues/1356

**Review by badb — APPROVED** [review]

**Comment by mfranciszkiewicz** [comment]
3 more tests are failing on macOS in comparison to develop (#1358):
https://gist.github.com/mfranciszkiewicz/57b8a328d98ee59f1149c1f9dddca0d6

**Comment by mfranciszkiewicz** [comment]
Local test issue:
```
WARNING  [golem.docker.task_thread           ] Task stderr:
Traceback (most recent call last):
  File "/golem/work/job.py", line 33, in <module>
    params.result_file)
  File "/golem/work/job.py", line 23, in run
    result_size)
  File "/golem/resources/code/computing.py", line 58, in run_dummy_task
    solution = find_pow(all_input, difficulty, result_size)
  File "/golem/resources/code/computing.py", line 34, in find_pow
    if check_pow(solution, input_data, difficulty):
  File "/golem/resources/code/computing.py", line 16, in check_pow
    sha.update(input_data.encode(encoding='UTF-8'))
UnicodeDecodeError: 'ascii' codec can't decode byte 0x8b in position 1: ordinal not in range(128)
```

**Review by mfranciszkiewicz — APPROVED** [review]
Please fix those (extra) failing tests on macOS.

**Comment by inexxt** [comment]
Waiting for https://github.com/golemfactory/golem/pull/1362 to be merged first

**Comment by mfranciszkiewicz** [comment]
Those 3 tests are still failing with #1362

**Comment by badb** [comment]
We're merging, @inexxt please fix the failing tests with separate PR.

**Comment by marmistrz** [comment]
@inexxt why are you using `jsonpickle` in your tests? We intend to remove the dependency altogether, `jsonpickle` was removed from the rest of the codebase in #1254

**Pull request merged**